vispy/visuals/colorbar.py

```diff
@@ -436,76 +314,125 @@ def _update(self):
-            text_x = x - _TEXT_PADDING_FACTOR * (halfw + self.border_width)
-            text_y = y
-            self._label.pos = text_x, text_y
-            self._label.rotation = -90
-
-            self._ticks[0].pos = text_x, y + halfh
-            self._ticks[1].pos = text_x, y - halfh
-            self._ticks[0].rotation = self.ticks[1].rotation = -90
-
-        else:
-            # raise an error since the orientation is now what was
-            # expected
-            raise ColorBarVisual._get_orientation_error(self._orientation)
-
-    def _prepare_transforms(self, view):
-        print("preparing transorms")
-        CompoundVisual._prepare_transforms(view)
-        view._update()
-
-    @staticmethod
-    def _get_orientation_error(orientation):
-        return ValueError("orientation must"
-                          " be 'horizontal' or 'vertical', "
-                          "not '%s'" % (orientation, ))
+    def _prepare_draw(self, view):
+        # print ("preparing draw")
+        self._update_positions()
 
     @staticmethod
     def _get_anchors(orientation):
         if orientation == "top":
             return "center", "bottom"
+
         elif orientation == "bottom":
             return "center", "top"
+
         elif orientation == "left":
             return "center", "bottom"
+
         else:  # orientation == "right"
             return "center", "top"
 
+    @staticmethod
+    def _calc_positions(center, halfdim, border_width,
+                        orientation, transforms):
+        """
+        Calculate the text centeritions given the ColorBar
+        parameters.
+
+        Note
+        ----
+        This is static because in principle, this
+        function does not need access to the state of the ColorBar
+        at all. It's a computation function that computes coordinate
+        transforms
+
+        Parameters
+        ----------
+        center: tuple (x, y)
+            Center of the ColorBar
+        halfdim: tuple (halfw, halfh)
+            Half of the dimensions measured from the center
+        border_width: float
+            Width of the border of the ColorBar
+        orientation: "top" | "bottom" | "left" | "right"
+            Position of the label with respect to the ColorBar
+        transforms: TransformSystem
+            the transforms of the ColorBar
+        """
+        (x, y) = center
+        (halfw, halfh) = halfdim
+
+        visual_to_doc = transforms.get_transform('visual', 'document')
+        doc_to_visual = transforms.get_transform('document', 'visual')
+
+        center_doc = visual_to_doc.map(np.array([x, y, 0, 1]))
+
+        half_axis_x = visual_to_doc.map(np.array([halfw, 0, 0, 1]))
+        half_axis_y = visual_to_doc.map(np.array([0, halfh, 0, 1]))
+
+        # -------------------
+        #               ^
+        #            half_axis_y
+        #               v
+        #               .(center)
+        # <-half_axis_x->
+        # --------------------
```

Comment:
Added an ASCII-picture describing the coordinate systems involved. Can remove if it's against guidelines